src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs

```diff
@@ -0,0 +1,253 @@
+        internal readonly MsQuicNativeMethods.StreamReceiveSetEnabledDelegate _streamReceiveSetEnabledDelegate;
+
+        internal readonly MsQuicNativeMethods.SetContextDelegate _setContextDelegate;
+        internal readonly MsQuicNativeMethods.GetContextDelegate _getContextDelegate;
+        internal readonly MsQuicNativeMethods.SetCallbackHandlerDelegate _setCallbackHandlerDelegate;
+
+        internal readonly MsQuicNativeMethods.SetParamDelegate SetParamDelegate;
+        internal readonly MsQuicNativeMethods.GetParamDelegate GetParamDelegate;
+
+        internal void RegistrationOpen(byte[] name)
+        {
+            MsQuicStatusException.ThrowIfFailed(_registrationOpenDelegate(name, out IntPtr ctx));
+            _registrationContext = ctx;
+        }
+
+        internal unsafe uint UnsafeSetParam(
+            IntPtr Handle,
+            uint Level,
+            uint Param,
+            MsQuicNativeMethods.QuicBuffer Buffer)
+        {
+            return SetParamDelegate(
+                Handle,
+                Level,
+                Param,
+                Buffer.Length,
+                Buffer.Buffer);
+        }
+
+        internal unsafe uint UnsafeGetParam(
+            IntPtr Handle,
+            uint Level,
+            uint Param,
+            ref MsQuicNativeMethods.QuicBuffer Buffer)
+        {
+            uint bufferLength = Buffer.Length;
+            byte* buf = Buffer.Buffer;
+            return GetParamDelegate(
+                Handle,
+                Level,
+                Param,
+                &bufferLength,
+                buf);
+        }
+
+        public async ValueTask<MsQuicSecurityConfig> CreateSecurityConfig(X509Certificate certificate)
+        {
+            MsQuicSecurityConfig secConfig = null;
+            var tcs = new TaskCompletionSource<object>();
+            uint secConfigCreateStatus = MsQuicConstants.InternalError;
+
+            uint status = _secConfigCreateDelegate(
+                _registrationContext,
+                (uint)QUIC_SEC_CONFIG_FLAG.CERT_CONTEXT,
+                certificate.Handle,
+                null,
+                IntPtr.Zero,
+                SecCfgCreateCallbackHandler);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            void SecCfgCreateCallbackHandler(
+                IntPtr context,
+                uint status,
+                IntPtr securityConfig)
+            {
+                secConfig = new MsQuicSecurityConfig(this, securityConfig);
+                secConfigCreateStatus = status;
+                tcs.SetResult(null);
+            }
+
+            await tcs.Task.ConfigureAwait(false);
+
+            MsQuicStatusException.ThrowIfFailed(secConfigCreateStatus);
+
+            return secConfig;
+        }
+
+        public IntPtr SessionOpen(byte[] alpn)
+        {
+            IntPtr sessionPtr = IntPtr.Zero;
+
+            uint status = _sessionOpenDelegate(
```

Comment:
Not completely sure on your guys style here, but all these places where you create this temporary variable and just pass it to `ThrowIfFailed` might be a bit cleaner like this:
```C#
MsQuicStatusException.ThrowIfFailed(
    _sessionOpenDelegate(
        _registrationContext,
        alpn,
        IntPtr.Zero,
        ref sessionPtr));
```